### PR TITLE
implement "avoid unnecessary setstate" rule

### PR DESCRIPTION
**Pull request by laptevw**
## Working on native implementations:  
  
* [x] implement "avoid unnecessary setstate" rule

## Screenshots
Снимок экрана 2023-08-28 в 11 46 06

    
## Quality  
  
* [x] No new warnings or errors using `fluter analyze`.  
* [x] All private methods starts with _ prefix.

### CODE REVIEW

File: example/test/avoid_unnecessary_setstate_test.dart [hunk 1]
```diff
@@ -0,0 +1,78 @@
+// ignore_for_file: avoid_global_state
```
Comment (solid-vovabeloded) [inline example/test/avoid_unnecessary_setstate_test.dart:1]: Why do we need to ignore it? Can we just fix it?

File: example/test/avoid_unnecessary_setstate_test.dart [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+/// Check unnecessary setstate fail
+/// `avoid_unnecessary_setstate`
+import 'package:flutter/material.dart';
+
+class MyWidget extends StatefulWidget {
```
Comment (solid-vovabeloded) [inline example/test/avoid_unnecessary_setstate_test.dart:7]: If you are using any code from the DCM package, please, add a license at the very beginning of the file.

File: example/test/avoid_unnecessary_setstate_test.dart [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      setState(() {
+        myString = "Hello";
+      });
+    }
+
+    // expect_lint: avoid_unnecessary_setstate
+    myStateUpdateMethod();
+
+    return ElevatedButton(
+      onPressed: () => myStateUpdateMethod(),
```
Comment (solid-vovabeloded) [inline example/test/avoid_unnecessary_setstate_test.dart:69]: NIT:
```suggestion
      onPressed: myStateUpdateMethod,
```

File: lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import 'package:analyzer/error/listener.dart';
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+import 'package:solid_lints/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart';
+import 'package:solid_lints/models/rule_config.dart';
+import 'package:solid_lints/models/solid_lint_rule.dart';
+
+/// A rule which warns when setState is called inside initState, didUpdateWidget
+/// or build methods and when it's called
+/// from a sync method that is called inside those methods.
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart:9]:
```suggestion
/// A rule which warns when setState is called inside initState, didUpdateWidget
/// or build methods and when it's called from a sync method that is called
/// inside those methods.
```

File: lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+class AvoidUnnecessarySetstateRule extends SolidLintRule {
+  /// The [LintCode] of this lint rule that represents
+  /// the error whether we return a widget.
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart:12]: Looks like a documentation comment from another class.

File: lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import 'package:analyzer/error/listener.dart';
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+import 'package:solid_lints/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart';
+import 'package:solid_lints/models/rule_config.dart';
+import 'package:solid_lints/models/solid_lint_rule.dart';
+
+/// A rule which warns when setState is called inside initState, didUpdateWidget
+/// or build methods and when it's called
+/// from a sync method that is called inside those methods.
+class AvoidUnnecessarySetstateRule extends SolidLintRule {
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart:10]:
```suggestion
class AvoidUnnecessarySetStateRule extends SolidLintRule {
```

File: lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart:1]: Please, rename `avoid_unnecessary_setstate_rule.dart` => `avoid_unnecessary_set_state_rule.dart`

File: lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    final visitor = AvoidUnnecessarySetStateVisitor();
+
+    context.registry.addClassDeclaration((node) {
+      visitor.visitClassDeclaration(node);
+      for (final element in visitor.setStateInvocations) {
+        reporter.reportErrorForNode(code, element);
+      }
+      for (final element in visitor.classMethodsInvocations) {
+        reporter.reportErrorForNode(code, element);
+      }
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_setstate_rule.dart:45]: Why do we need two different lists here if we are not distinguishing them when showing an error? Does it make sense to have only one list of invocations?

File: lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    visitedRestMethods[name] = hasSetState;
+
+    return hasSetState;
+  }
+}
+
+/// AST Visitor which finds all setState invocations and checks if they are
+/// necessary
+class _MethodVisitor extends RecursiveAstVisitor<void> {
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart:120]: I'd suggest moving this class into a separate file and giving it a more meaningful name.

File: lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  final Set<String> classMethodsNames;
+  final Iterable<FunctionBody> bodies;
+
+  final _setStateInvocations = <MethodInvocation>[];
+  final _classMethodsInvocations = <MethodInvocation>[];
+
+  Iterable<MethodInvocation> get setStateInvocations => _setStateInvocations;
+  Iterable<MethodInvocation> get classMethodsInvocations =>
+      _classMethodsInvocations;
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart:129]: Please, apply the correct formatting.

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Well done, @laptevw! I've added a couple of suggestions here - please, take a look.

File: lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart (hunk 9 above)
Comment (laptevw) [reply lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart:129]: This line is automatically formatted by Android Studio according to the maxLineLength - 80 rule.
Is it correct?

Comment (solid-vovabeloded) [reply lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart]: I mean we should have an empty line between getters. Android Studio should automatically apply this formatting.

File: lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_set_state_rule.dart [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import 'package:solid_lints/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_set_state_visitor.dart';
+import 'package:solid_lints/models/rule_config.dart';
+import 'package:solid_lints/models/solid_lint_rule.dart';
+
+/// A rule which warns when setState is called inside initState, didUpdateWidget
+/// or build methods and when it's called from a sync method that is called
+/// inside those methods.
+class AvoidUnnecessarySetStateRule extends SolidLintRule {
+  /// The [LintCode] of this lint rule that represents
+  /// the error whether we use setState in inappropriate way.
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/avoid_unnecessary_set_state_rule.dart:12]: I think it is not a `LintCode`, but a lint name, so we should not use a `LintCode` reference in the documentation.

File: lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_set_state_method_visitor.dart [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import 'package:analyzer/dart/ast/visitor.dart';
+
+/// AST Visitor which finds all setState invocations and checks if they are
+/// necessary
+class AvoidUnnecessarySetStateMethodVisitor extends RecursiveAstVisitor<void> {
+  final Set<String> _classMethodsNames;
+  final Iterable<FunctionBody> _bodies;
+
+  final _setStateInvocations = <MethodInvocation>[];
+  final _classMethodsInvocations = <MethodInvocation>[];
```
Comment (solid-vovabeloded) [inline lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_set_state_method_visitor.dart:34]: Do we actually need 2 separate lists here?

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Great job, @laptevw! I've added a couple of minor suggestions here, but overall, the code looks much better!

**Pull request merged**